Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm

```diff
@@ -107,13 +110,62 @@
 
     URL modifiedRequestURL { modifiedRequestURLString.get() };
     if (!modifiedRequestURL.isValid()) {
-        LOG(ContentFiltering, "NetworkExtensionContentFilter failed to convert modified URL string %@ to a  URL.\n", modifiedRequestURLString.get());
+        LOG(ContentFiltering, "NetworkExtensionContentFilter failed to convert modified URL string %@ to a URL.\n", modifiedRequestURLString.get());
         return;
     }
 
     request.setURL(WTFMove(modifiedRequestURL));
 }
 
+void NetworkExtensionContentFilter::willSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&, State)>&& completionHandler)
+{
+    ASSERT(!request.isNull());
+    if (!request.url().protocolIsInHTTPFamily() || !enabled()) {
+        m_state = State::Allowed;
+        completionHandler(WTFMove(request), m_state);
+        return;
+    }
+
+    initialize();
+
+    if (!redirectResponse.isNull()) {
+        responseReceived(redirectResponse);
+        if (!needsMoreData()) {
+            completionHandler(WTFMove(request), m_state);
+            return;
+        }
+    }
+
+    RetainPtr nsRequest = request.protectedNSURLRequest(DoNotUpdateHTTPBody);
+
+    auto blockPtr = makeBlockPtr([protectedThis = Ref { *this }, request = WTFMove(request), completionHandler = WTFMove(completionHandler)](NEFilterSourceStatus status, NSDictionary *decisionInfo) mutable {
+        RetainPtr<NSString> modifiedRequestURLString;
+        if (decisionInfo)
+            modifiedRequestURLString = checked_objc_cast<NSString>(decisionInfo[NEFilterSourceOptionsRedirectURL]);
+
+        RunLoop::mainSingleton().dispatch([status, decisionInfo = RetainPtr { decisionInfo }, modifiedRequestURLString = WTFMove(modifiedRequestURLString), protectedThis = WTFMove(protectedThis), request = WTFMove(request), completionHandler = WTFMove(completionHandler)] mutable {
+            protectedThis->handleDecision(status, replacementDataFromDecisionInfo(decisionInfo.get()));
+
+            if (!modifiedRequestURLString) {
+                completionHandler(WTFMove(request), protectedThis->m_state);
+                return;
+            }
+
+            URL modifiedRequestURL { modifiedRequestURLString.get() };
+            if (!modifiedRequestURL.isValid()) {
+                LOG(ContentFiltering, "NetworkExtensionContentFilter failed to convert modified URL string %@ to a URL.\n", modifiedRequestURLString.get());
```

Comment:
Can we do a RELEASE_LOG_ERROR here? It feels bad to silently ignore the URL in this case.